### PR TITLE
Updated the geo normalizers to accept valid hashes

### PR DESCRIPTION
**Pull request by ivakegg**

### CODE REVIEW

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java [hunk 1]
```diff
@@ -49,7 +50,15 @@
...
-        return normalizeDelegateType(createDatawaveGeometry((G) parseGeometry(geoString)));
+        try {
+            return normalizeDelegateType(createDatawaveGeometry((G) parseGeometry(geoString)));
+        } catch (Exception e) {
+            if (validHash(geoString)) {
+                return geoString;
+            }
+            Throwables.propagateIfPossible(e, IllegalArgumentException.class);
+        }
+        throw new IllegalArgumentException("Unable to normalize geo string " + geoString);
```
Comment (apmoriarty) [inline src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:61]: Does it make sense to check for a valid hash first, and only attempt to normalize if an invalid hash is found? That would save all the work spent to build a stack trace

File: src/main/java/datawave/data/normalizer/GeometryNormalizer.java [hunk 2]
```diff
@@ -49,7 +52,16 @@ protected Index getIndex() {
     
     @Override
     public List<String> normalizeToMany(String geoString) throws IllegalArgumentException {
-        return normalizeDelegateTypeToMany(createDatawaveGeometry(parseGeometry(geoString)));
+        try {
+            return normalizeDelegateTypeToMany(createDatawaveGeometry(parseGeometry(geoString)));
+        } catch (Exception e) {
+            // perhaps this is a geo hash instead
+            if (validHash(geoString)) {
```
Comment (apmoriarty) [inline src/main/java/datawave/data/normalizer/GeometryNormalizer.java:59]: same comment as before, re: validhash first, then normalize if invalid

**Review by apmoriarty — CHANGES_REQUESTED** [review]

**Review by apmoriarty — DISMISSED** [review]
LGTM

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java [hunk 3]
```diff
@@ -175,4 +178,48 @@ private String adjustHexRangeBigInteger(String hexValue, String format, boolean
...
+        return Long.parseLong(value.substring(2), 16);
+    }
+    
+    public boolean validTier(short tier) {
+        return tier >= 0 && tier <= 0x1f;
+    }
+    
+    public boolean validLength(short tier, String value) {
+        // determine the length of the position in hex characters
+        long posLen = Math.round(Math.ceil((double) tier / 4)) * 2;
```
Comment (jwomeara) [inline src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:216]: Why are you rounding?

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java [hunk 4]
```diff
@@ -175,4 +178,48 @@ private String adjustHexRangeBigInteger(String hexValue, String format, boolean
...
+        // length is the tier length plus the position length
+        return value.length() == (2 + posLen);
+    }
+    
+    public boolean validPosition(short tier, long value) {
+        return value >= 0 && value < Math.round(Math.pow(2.0d, 2.0d * tier));
```
Comment (jwomeara) [inline src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:222]: Why are you rounding?

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java [hunk 5]
```diff
@@ -175,4 +178,48 @@ private String adjustHexRangeBigInteger(String hexValue, String format, boolean
...
+    
+    public long getPosition(String value) {
+        if (value.length() == 2) {
+            return 0;
+        }
+        return Long.parseLong(value.substring(2), 16);
+    }
+    
+    public boolean validTier(short tier) {
+        return tier >= 0 && tier <= 0x1f;
```
Comment (jwomeara) [inline src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:211]: This is fine, but if we ever decide to change the 'indexStrategy' used by PointNormalizer and GeometryNormalizer, we may need to update these checks to behave differently.  Might be worth just adding a comment here, or on the line where the indexStrategy is defined so we don't forget.

**Review by jwomeara — CHANGES_REQUESTED** [review]

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 3 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:216]: Takes care of the double conversions and ensures I am getting a long

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 4 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:222]: Takes care of the double conversions and ensures I am getting a long

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 3 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:216]: I have seen too many times where I get a double like 1.99999999999999999 which was supposed to be a 2.00000000000000

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 5 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java:211]: oh?  I did not realized this.  I will add a very large bold comment in this regard

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 3 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java]: ok, removed the round and changed to a cast instead

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 4 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java]: changed to a cast

**Review by jwomeara — DISMISSED** [review]

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 4 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java]: changed to use bit shifts which is more efficient and guaranteed to be exact

File: src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java (hunk 3 above)
Comment (ivakegg) [reply src/main/java/datawave/data/normalizer/AbstractGeometryNormalizer.java]: changed to use bit shifts which is more efficient and guaranteed to be exact

**Pull request merged**